File: src/main/java/cn/edu/tsinghua/sdfs/client/handler/FileSendHandler.java
package cn.edu.tsinghua.sdfs.client.handler;

import cn.edu.tsinghua.sdfs.codec.Codec;
import cn.edu.tsinghua.sdfs.protocol.FilePacket;
import io.netty.buffer.ByteBuf;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.stream.ChunkedFile;

import java.io.File;

public class FileSendHandler extends ChannelInboundHandlerAdapter {

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        ByteBuf byteBuf = (ByteBuf) msg;
        FilePacket filePacket = (FilePacket) Codec.INSTANCE.decode(byteBuf);
        System.out.println("prepared send: " + filePacket.getFile().getName());

        Channel channel = ctx.channel();
        channel.writeAndFlush(new ChunkedFile(new File(filePacket.getFile().getName())));

    }
}

File: src/main/java/cn/edu/tsinghua/sdfs/client/handler/FileSendHandler.kt
package cn.edu.tsinghua.sdfs.client.handler

import cn.edu.tsinghua.sdfs.codec.Codec
import cn.edu.tsinghua.sdfs.protocol.packet.impl.CreateRequest
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.stream.ChunkedFile
import java.io.File

class FileSendHandler : ChannelInboundHandlerAdapter() {

    @Throws(Exception::class)
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        val byteBuf = msg as ByteBuf
        val packet = Codec.INSTANCE.decode(byteBuf)

        when {
            packet is CreateRequest && packet.localFile.isNotEmpty() -> {
                val channel = ctx.channel()
                channel.writeAndFlush(ChunkedFile(File(packet.localFile)))
            }
        }

    }
}
